peak/ir.py

```diff
@@ -0,0 +1,53 @@
+from collections import namedtuple
+import typing as tp
+from hwtypes.adt import Product
+from hwtypes import AbstractBitVector, AbstractBit, BitVector, Bit
+from .peak import Peak, name_outputs
+import itertools as it
+
+class IR:
+    def __init__(self):
+        #Stores a list of instructions (name : family_closure)
+        self.instructions = {}
+
+    def add_instruction(self, name, peak_fc : tp.Callable):
+        if name in self.instructions:
+            raise ValueError(f"{name} is already an existing instruction")
+        self.instructions[name] = peak_fc
+
+    #fun should have the form def fun(family,*args)
+    def add_peak_instruction(self, name : str, input_interface : Product, output_interface : Product, fun : tp.Callable):
+        #Assuming for now that abstract bitvectors are used in the interfaces
+        inputs = input_interface.field_dict
+        outputs = output_interface.field_dict
+
+        def ts(n):
+            assert n>0
+            return "".join(["    " for _ in range(n)])
+        t_to_tname = {}
+        idx = 0
+        for t in it.chain(inputs.values(),outputs.values()):
+            if not t in t_to_tname:
+                t_to_tname[t] = f"t{idx}"
+                idx +=1
+        class_src = f"def peak_fc(family):\n"
```

Comment:
build `class_src` as list instead appending with `+=`.
Also remove `\n` from the end of each line.
before execing add the line `class_src = '\n'.join(class_src)`